### PR TITLE
Added virtual function OnStateSet

### PR DESCRIPTION
**Pull request by chris5287**
Added virtual function OnStateSet to allow a subscriber to perform work before StateHasChanged is called (a bit like OnParametersSet).

### CODE REVIEW

File: Source/BlazorState/Components/BlazorStateComponent.cs [hunk 1]
```diff
@@ -42,10 +42,19 @@ public BlazorStateComponent()
...
+    /// </summary>
+    public virtual void OnStateSet() {}
+
     /// <summary>
     /// Exposes StateHasChanged
     /// </summary>
-    public void ReRender() => base.InvokeAsync(StateHasChanged);
+    public void ReRender()
+    {
+      base.InvokeAsync(OnStateSet);
```
Comment (chris5287) [inline Source/BlazorState/Components/BlazorStateComponent.cs]: Wasn't sure if I should call OnStateSet directly or via InvokeAsync?

**Comment by pablopioli** [comment]
This addition sounds interesting, please merge.

Would simplify a couple of edge cases I have found.

**Comment by StevenTCramer** [comment]
@pablopioli  could you please give me example of the edge cases where this would be of benefit?

See Pete's Docs for nice overview of Blazor component lifecycle.
https://blazor-university.com/components/component-lifecycles/

**Comment by pablopioli** [comment]
I had a table and a chart (JS based) of the table data.

In a handler I can update the state and trigger a table refresh. But the chart needs to be manually redrawn in the handler. This leads to a lot of coupling.

If I can be notified of the state update I can trigger updates on components that don´t work natively in Blazor.

**Comment by StevenTCramer** [comment]
Pablo, have you looked at how the pipeline works. One can fire an event and handle this event from anywhere. This would be more in line with my objectives of the repo.

**Comment by pablopioli** [comment]
I have looked at INotification and the hooks provided by Mediatr, the problem is that they assume a non-UI object. It is difficult to update the UI directly from there. 

As you suggested I used a Domain Events style approach to solve my problem. I send events from the handler, which Blazor components subscribe to. It works, but is more error prone and you end having two pipelines of processing (BlazorState and the Domain Events). My app was small and worked fine, but in a bigger app I think it will be confusing.

Don't worry, I understand your point of view, I just have to warn you that the code will be harder to think of when you need to do non trivial JS interop (at least from my point of view). If someday I need to do that I would probably do a fork of your code and incorporate this changes.

**Comment by StevenTCramer** [comment]
With the latest updates in TimeWarp.State I hope this is no longer desired.

**Pull request closed without merging**